### ISSUE TITLE
Question about the adaptive_dilated_conv2d

### ISSUE THREAD

**Issue by donggoing** [BIGKnight/PytorchCudaOpExtension#1]
I am try to using this adaptive_dilated_conv2d operator to implement module can apply different specific dialated rates for different pixels. But it seem doesn't work. When i use it with fixed dialated rate(all set to 1), i suppose it should work just like conv2d, but it doesn't. The loss using adaptive_dilated_conv2d is mounting while falls normally using conv2d. I'm wondering why.

**Comment by BIGKnight** [comment]
well, thanks for your attention, i'm not very sure about the specific reasons since it has been a long time after my last updates about this repo(torch has been updated for several versions...). One possible reason, i guess, is that the learnable parameter [self.rates](https://github.com/BIGKnight/PytorchCudaOpExtension/blob/0255d0f294ce4911015a2c65abb82c3648ee42d8/adaptive_dilated_conv2d/adaptive_dilated_conv2d_wrapper.py#L42) have not been fixed, so during the updates, it still gots some updates?  Did you fixed this part?

**Comment by donggoing** [comment]
Actually instead of using self.rates, I send a specific rate map to do the calculation. Like that:
```python
class AdaptiveDilatedConv2dFunction(Function):
    @staticmethod
    def forward(ctx, *args):
        if len(args) != 6:
            print("wrong input parameters number, check the input")
            return
        input = args[0]
        weights = args[1]
        rate_map = args[2]
        bias = args[3]
        ctx.stride_h = args[4]
        ctx.stride_w = args[5]
        output = adaptive_dilated_conv2d.forward(input, weights, rate_map, bias, ctx.stride_h, ctx.stride_w)
        ctx.save_for_backward(input, weights, rate_map, bias)
        return output

    @staticmethod
    def backward(ctx, *grad_outputs):
        if len(grad_outputs) != 1:
            print("Wrong output number, check your output")
            return
        input, weights, rate_map, bias = ctx.saved_tensors
        grad = grad_outputs[0].clone()
        grad_input, grad_weight, grad_rate_map, grad_bias = adaptive_dilated_conv2d.backward(input, weights, rate_map, bias, grad, ctx.stride_h, ctx.stride_w)
        return grad_input, grad_weight, grad_rate_map, grad_bias, None, None


class AdaptiveDilatedConv2dLayer(Module):
    def __init__(self, in_channels, out_channels, kernel_size, stride_h, stride_w):
        super(AdaptiveDilatedConv2dLayer, self).__init__()
        self.kernel_size = kernel_size
        self.pad = (kernel_size // 2)
        self.stride_h = stride_h
        self.stride_w = stride_w
        self.weight = nn.Parameter(torch.zeros(out_channels, in_channels, kernel_size, kernel_size, dtype=torch.float32))
        # self.rates = nn.Parameter(torch.ones(1, 1, img_scale, img_scale,  dtype=torch.float32) * dilation)
        self.bias = nn.Parameter(torch.zeros(out_channels, dtype=torch.float32))
        nn.init.xavier_uniform_(self.weight, gain=1)
        
    def forward(self, inputs, rates):
        inputs = torch.nn.functional.pad(inputs, [self.pad, self.pad, self.pad, self.pad ])
        # h_in, w_in = inputs.shape[2:4]
        # h_out = (h_in - (1 * (self.kernel_size - 1) + 1)) // self.stride_h + 1 # already add padding in former op, thus it do not worry about the division
        # w_out = (w_in - (1 * (self.kernel_size - 1) + 1)) // self.stride_w + 1
        # rate_map = nn.functional.interpolate(rates, size=[h_out, w_out], mode='nearest')
        return AdaptiveDilatedConv2dFunction.apply(inputs, self.weight, rates, self.bias, self.stride_h, self.stride_w)
```

And for operator code, i just change `THCState_getCurrentStream(state)` to `c10::cuda::getCurrentCUDAStream()` and `type().is_cuda()` to `is_cuda()`.

**Comment by BIGKnight** [comment]
emmm, i'm sorry i dont know the reason, maybe there is some problem in the cuda code. but currently, i am not very willing to re-check this three-years-old code now...😂 (actually, this repo was just a recording of my cuda learning, I did not carry on very careful code check though it worked for me on a small dataset, and I did not need to write cuda code or implement some custom operators recently so I nearly forget the cuda related knowledge now). 

However, the code of .cu part was written with the help of the official repo of [deformation convolution](https://github.com/msracver/Deformable-ConvNets). they used the MXNET deep learning framework while I applied it in Torch, but the code would be similar with respect to the .cu file(CUDA C language). i suggest you to refer theirs directly. I hope this could help😂.
